### DISCUSSION TITLE
jace Emulator support for Apple2-IO-RPi

### DISCUSSION THREAD

**Discussion by hanshuebner** [tjboldt/Apple2-IO-RPi#93] [General]
Hi,

I coded up some initial support for [emulating the Apple2-IO-RPi](https://github.com/hanshuebner/jace/blob/master/src/main/java/jace/hardware/CardApple2IORPi.java) in the [jace emulator](https://github.com/hanshuebner/jace/) so that I could work on my [Lisp based server](https://github.com/hanshuebner/a2-server) while traveling.  It uses UDP to send and receive bytes between the emulator and the server.  If anyone is interested in adding support to the Go based server and needs help with this or documentation, let me know.

The hardware emulation is not complete, in particular paging support for the ROM space is missing - I don't yet completely understand how it works, but my goal is to fully support it so that all hardware features can be tested in the emulator.

-Hans

**Comment by tjboldt** [comment]
The firmware paging in hardware works by using bits 4 and 5 of the 74LVC374 latch and routing them A11 and A12 of the EEPROM. Note that not only is their paging but the firmware is different per slot which is why it uses A0..A10 instead of A0..A7 for the 256 byte space. When you change bits 4 and 5 writing to OutputFlags = $C087+SLOT*$10, it changes the firmware page. Let's take slot 7 with OutputFlags set to $0F for an example... The 256 bytes of firmware come from $1C00-$1CFF in the EEPROM. Writing $3F to address $C0F7 would make the firmware come from $1F00-$1FFF in the EEPROM.
